certs/service_test.go

```diff
@@ -33,217 +35,300 @@ const (
+				PrivateKeyType: "",
+				Serial:         "",
+				Expire:         0,
+			},
+			identifyRes:  &magistrala.IdentityRes{Id: validID},
+			identifyErr:  nil,
+			thingErr:     nil,
+			issueCertErr: nil,
+			err:          nil,
 		},
 		{
 			desc:    "issue new cert for non existing thing id",
 			token:   token,
 			thingID: "2",
 			ttl:     ttl,
-			err:     certs.ErrFailedCertCreation,
+			pki: pki.Cert{
+				ClientCert:     "",
+				IssuingCA:      "",
+				CAChain:        []string{},
+				ClientKey:      "",
+				PrivateKeyType: "",
+				Serial:         "",
+				Expire:         0,
+			},
+			identifyRes:  &magistrala.IdentityRes{Id: validID},
+			identifyErr:  svcerr.ErrAuthentication,
+			thingErr:     errors.NewSDKError(certs.ErrFailedCertCreation),
+			issueCertErr: certs.ErrFailedCertCreation,
+			err:          svcerr.ErrAuthentication,
 		},
 		{
 			desc:    "issue new cert for non existing thing id",
 			token:   invalid,
 			thingID: thingID,
 			ttl:     ttl,
-			err:     svcerr.ErrAuthentication,
+			pki: pki.Cert{
+				ClientCert:     "",
+				IssuingCA:      "",
+				CAChain:        []string{},
+				ClientKey:      "",
+				PrivateKeyType: "",
+				Serial:         "",
+				Expire:         0,
+			},
+			identifyRes:  &magistrala.IdentityRes{Id: validID},
+			identifyErr:  svcerr.ErrAuthentication,
+			thingErr:     errors.NewSDKError(certs.ErrFailedCertCreation),
+			issueCertErr: certs.ErrFailedCertCreation,
+			err:          svcerr.ErrAuthentication,
 		},
 	}
 
 	for _, tc := range cases {
-		repoCall := auth.On("Identify", mock.Anything, &magistrala.IdentityReq{Token: tc.token}).Return(&magistrala.IdentityRes{Id: validID}, nil)
-		repoCall1 := auth.On("Authorize", mock.Anything, mock.Anything).Return(&magistrala.AuthorizeRes{Authorized: true}, tc.err)
-		repoCall2 := sdk.On("Thing", mock.Anything, mock.Anything).Return(mgsdk.Thing{ID: tc.thingID, Credentials: mgsdk.Credentials{Secret: thingKey}}, errors.NewSDKError(tc.err))
+		authCall := auth.On("Identify", context.Background(), &magistrala.IdentityReq{Token: tc.token}).Return(tc.identifyRes, tc.identifyErr)
+		sdkCall := sdk.On("Thing", tc.thingID, tc.token).Return(mgsdk.Thing{ID: tc.thingID, Credentials: mgsdk.Credentials{Secret: thingKey}}, tc.thingErr)
+		agentCall := agent.On("IssueCert", thingKey, tc.ttl).Return(tc.pki, tc.issueCertErr)
+		repoCall := repo.On("Save", context.Background(), mock.Anything).Return("", tc.err)
+
 		c, err := svc.IssueCert(context.Background(), tc.token, tc.thingID, tc.ttl)
 		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
 		cert, _ := certs.ReadCert([]byte(c.ClientCert))
 		if cert != nil {
 			assert.True(t, strings.Contains(cert.Subject.CommonName, thingKey), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, thingKey, cert.Subject.CommonName))
 		}
+		authCall.Unset()
+		sdkCall.Unset()
+		agentCall.Unset()
 		repoCall.Unset()
-		repoCall1.Unset()
-		repoCall2.Unset()
 	}
 }
```

Comment:
Something like this will work for all calls
```suggestion
func TestIssueCert(t *testing.T) {
	svc, repo, agent, auth, sdk := newService(t)
	cases := []struct {
		token       string
		desc        string
		thingID     string
		ttl         string
		key         string
		pki         pki.Cert
		identifyRes *magistrala.IdentityRes
		authErr     error
		sdkErr      errors.SDKError
		agentErr    error
		repoErr     error
		err         error
	}{
		{
			desc:    "issue new cert without any errors",
			token:   token,
			thingID: thingID,
			ttl:     ttl,
			pki: pki.Cert{
				ClientCert:     "",
				IssuingCA:      "",
				CAChain:        []string{},
				ClientKey:      "",
				PrivateKeyType: "",
				Serial:         "",
				Expire:         0,
			},
			identifyRes: &magistrala.IdentityRes{Id: validID},
		},
		{
			desc:    "issue new cert for with invalid token",
			token:   invalid,
			thingID: thingID,
			ttl:     ttl,
			pki: pki.Cert{
				ClientCert:     "",
				IssuingCA:      "",
				CAChain:        []string{},
				ClientKey:      "",
				PrivateKeyType: "",
				Serial:         "",
				Expire:         0,
			},
			identifyRes: &magistrala.IdentityRes{Id: validID},
			authErr:     svcerr.ErrAuthentication,
			err:         svcerr.ErrAuthentication,
		},
		{
			desc:    "issue new cert for non existing thing id",
			token:   token,
			thingID: "2",
			ttl:     ttl,
			pki: pki.Cert{
				ClientCert:     "",
				IssuingCA:      "",
				CAChain:        []string{},
				ClientKey:      "",
				PrivateKeyType: "",
				Serial:         "",
				Expire:         0,
			},
			identifyRes: &magistrala.IdentityRes{Id: validID},
			sdkErr:      errors.NewSDKError(errors.ErrMalformedEntity),
			err:         certs.ErrFailedCertCreation,
		},
		{
			desc:    "failed to issue cert by agent",
			token:   token,
			thingID: "2",
			ttl:     ttl,
			pki: pki.Cert{
				ClientCert:     "",
				IssuingCA:      "",
				CAChain:        []string{},
				ClientKey:      "",
				PrivateKeyType: "",
				Serial:         "",
				Expire:         0,
			},
			identifyRes: &magistrala.IdentityRes{Id: validID},
			agentErr:    pki.ErrFailedCertCreation,
			err:         certs.ErrFailedCertCreation,
		},
		{
			desc:    "failed to save in repo",
			token:   token,
			thingID: "2",
			ttl:     ttl,
			pki: pki.Cert{
				ClientCert:     "",
				IssuingCA:      "",
				CAChain:        []string{},
				ClientKey:      "",
				PrivateKeyType: "",
				Serial:         "",
				Expire:         0,
			},
			identifyRes: &magistrala.IdentityRes{Id: validID},
			repoErr:     certs.ErrMissingCerts,
			err:         certs.ErrMissingCerts,
		},
	}

	for _, tc := range cases {
		authCall := auth.On("Identify", context.Background(), &magistrala.IdentityReq{Token: tc.token}).Return(tc.identifyRes, tc.authErr)
		sdkCall := sdk.On("Thing", tc.thingID, tc.token).Return(mgsdk.Thing{ID: tc.thingID, Credentials: mgsdk.Credentials{Secret: thingKey}}, tc.sdkErr)
		agentCall := agent.On("IssueCert", thingKey, tc.ttl).Return(tc.pki, tc.agentErr)
		repoCall := repo.On("Save", context.Background(), mock.Anything).Return("", tc.repoErr)

		c, err := svc.IssueCert(context.Background(), tc.token, tc.thingID, tc.ttl)
		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))

		cert, _ := certs.ReadCert([]byte(c.ClientCert))
		if cert != nil {
			assert.True(t, strings.Contains(cert.Subject.CommonName, thingKey), fmt.Sprintf("%s: expected %s got %s\n", tc.desc, thingKey, cert.Subject.CommonName))
		}

		if tc.err == nil {
			ok := authCall.Parent.AssertCalled(t, "Identify", context.Background(), &magistrala.IdentityReq{Token: tc.token})
			assert.True(t, ok, fmt.Sprintf("Identify was not called on %s", tc.desc))

			ok = sdkCall.Parent.AssertCalled(t, "Thing", tc.thingID, tc.token)
			assert.True(t, ok, fmt.Sprintf("Thing was not called on %s", tc.desc))

			ok = agentCall.Parent.AssertCalled(t, "IssueCert", thingKey, tc.ttl)
			assert.True(t, ok, fmt.Sprintf("IssueCert was not called on %s", tc.desc))

			ok = repoCall.Parent.AssertCalled(t, "Save", context.Background(), mock.Anything)
			assert.True(t, ok, fmt.Sprintf("Save was not called on %s", tc.desc))
		}

		authCall.Unset()
		sdkCall.Unset()
		agentCall.Unset()
		repoCall.Unset()
	}
}
```